Improve TOTP field

DL6ER opened the pull request:
# What does this implement/fix?

Fix misaligned TOTP field icon, make password field larger when TOTP is enabled and fix auto-submission to only fire when the user has already put in something into the password field.

# Screenshots

### TOTP not enabled, before and after
![Screenshot from 2023-10-30 08-18-27](https://github.com/pi-hole/web/assets/16748619/5878d796-95df-4c6b-9d84-439d54eb7eef)

### TOTP enabled, `development-v6`
(misaligned icon, small password field)
![Screenshot from 2023-10-30 08-19-10](https://github.com/pi-hole/web/assets/16748619/706b9bff-ddb7-4fae-911c-dbae688b8e34)

### TOTP enabled, this PR
![Screenshot from 2023-10-30 08-27-49](https://github.com/pi-hole/web/assets/16748619/32d132ff-4bbb-42a3-b758-c17f98fd7e5d)

**Related issue or feature (if applicable):** N/A

**Pull request in [docs](https://github.com/pi-hole/docs) with documentation (if applicable):** N/A

---
**By submitting this pull request, I confirm the following:** 

1. I have read and understood the [contributors guide](https://docs.pi-hole.net/guides/github/contributing/), as well as this entire template. I understand which branch to base my commits and Pull Requests against. 
2. I have commented my proposed changes within the code.
3. I am willing to help maintain this change if there are issues with it later.
4. It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
5. I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

## Checklist:

- [x] The code change is tested and works locally.
- [x] I based my code and PRs against the repositories `developmental` branch.
- [x] I [signed off](https://docs.pi-hole.net/guides/github/how-to-signoff/) all commits. Pi-hole enforces the [DCO](https://docs.pi-hole.net/guides/github/dco/) for all contributions
- [x] I [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) all my commits. Pi-hole requires signatures to verify authorship
- [x] I have read the above and my PR is ready for review.

DL6ER committed:
Fix position of TOTP token

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Only submit form after entering TOTP token when there is something in the password field as well

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER committed:
Follow best-practice rules for login pages from https://web.dev/articles/sign-in-form-best-practices

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
https://web.dev/articles/sign-in-form-best-practices

This adds a "show password" button:
![image](https://github.com/pi-hole/web/assets/16748619/b72e2728-52b6-4e75-8b18-8bace8c37e18)

yubiuser reviewed (commented):
I think we should add a bit of space between the `https` warning and `show password`
![Screenshot at 2023-10-30 20-40-06](https://github.com/pi-hole/web/assets/26622301/066d1232-0801-44ff-a2db-9a9fa5035f54)

rdwebdesign commented:
Do we need the key icon?
Can we use an eye icon as "show password" button?

Something like this:
![password_reveal](https://github.com/pi-hole/web/assets/1385443/0ea3ab94-80e4-48bd-bc07-11005f001433)

DL6ER commented:
That's a very good solution, if you have this already at hand ;-) Feel free to push ti this branch

rdwebdesign commented:
I will do it in a few minutes.

rdwebdesign committed:
Add an inline button to show/hide the password

using input-group and input-group-btn classes.
Also remove the key icon from the password field.

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign committed:
Function to toggle password visibility (using jQuery) and icon

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

rdwebdesign committed:
Remove unused style

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

yubiuser reviewed (commented):
The click/hover animation (color) changes stays after click until one clicks somewhere else.
![Peek 2023-11-03 22-13](https://github.com/pi-hole/web/assets/26622301/26669362-af0c-4698-9cef-e35aa895c469)

I think this should be a hover only.

yubiuser reviewed (commented):
Entering the password first and submit correct TOTP afterwards shows "Wrong password" briefly before auto-login

![Peek 2023-11-03 22-20](https://github.com/pi-hole/web/assets/26622301/a6a11b1e-05cd-4ebf-a3de-2cab132dbcba)

yubiuser commented on login.lp:
```diff
@@ -43,21 +43,20 @@ mg.include('scripts/pi-hole/lua/header.lp','r')
-                        <label for="logincookie">Keep me logged in (uses cookie)</label>
-                    </div> -->
+                    <div class="form-group has-feedback hidden" id="totp_input">
+                        <input type="text" id="totp" size="6" maxlen="6" class="form-control totp_token" placeholder="123456" value="" spellcheck="false" autofocus>
```
```suggestion
                        <input type="text" id="totp" size="6" maxlen="6" class="form-control totp_token" placeholder="123456" value="" spellcheck="false" autocomplete="off" autofocus>
```

Not sure if this is the right command, but there should be no (sensless) autofill/complete on the TOTP input field

yubiuser reviewed (commented):
There is no distinction between entering a wrong password or a wrong TOTP. It will always complain about "Wrong password" and autofocus on the password field - even when the TOTP was wrong.

rdwebdesign commented:
This is the `:focus` style.
Every button has the same effect, but usually the focus changes to some other element after the click.

We can change the focus to the password field, like this:
![button_focus](https://github.com/pi-hole/web/assets/1385443/6026df1b-a0b9-4fbc-8339-c8574fa4ea07)

rdwebdesign committed:
Set the focus to password field after clicking on the button

Signed-off-by: RD WebDesign <github@rdwebdesign.com.br>

DL6ER commented:
> Entering the password first and submit correct TOTP afterwards shows "Wrong password" briefly before auto-login

I am not able to reproduce this and looking again at the Javascript code, it also isn't immediately obvious to me how this could be happening. Could you try again and screenshot the `Network` tab frpm your browser for me with additional screenshots of the reply contents?

Please make sure to have "Persist Logs" enabled for them to "survive" the login process.

![image](https://github.com/pi-hole/web/assets/16748619/bc062b1c-4fb6-4448-a102-c32ced4d5e26)

---

> There is no distinction between entering a wrong password or a wrong TOTP. It will always complain about "Wrong password" and autofocus on the password field - even when the TOTP was wrong.

I will change this to be clearer.

DL6ER committed:
Show dedicated error mesage on wrong 2FA token

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
## Wrong password

![Screenshot from 2023-11-04 07-16-01](https://github.com/pi-hole/web/assets/16748619/7af9846e-faaa-470c-ad15-a0b6cdcd9b69)

## Correct password but wrong 2FA token

![Screenshot from 2023-11-04 07-16-17](https://github.com/pi-hole/web/assets/16748619/0f56162b-96b2-4cdd-bf53-a7da04b91aa4)

yubiuser commented:
If you think letting users know if they entered a wrong password or a wrong TOTP is a security issue, I'm fine with leaving it as it is. I know this discussion, but it feels a bit like "security by obfuscation". If the second factor is so weak that it can be brute-forced within 30 seconds (before a new TOTP is generated), the second factor is broken anyway.

DL6ER commented:
Yeah, since online-bruteforcing is prevented by FTL and when they can get physical/`root` access to either the Pi-hole or the device having the Authenticator app they'll trivially have the TOTP secret, I am fine with showing the user whether it was the password or the token that was wrong. Seem more user-friendly to me.

yubiuser commented:
> Entering the password first and submit correct TOTP afterwards shows "Wrong password" briefly before auto-login

I'm not seeing this today, but there are two calls to `/auth` with the second one failing which might be the reason I saw it yesterday. Disabling 2FA, only one `/auth` request is logged.

![Screenshot at 2023-11-05 20-47-49](https://github.com/pi-hole/web/assets/26622301/22ce23be-6664-4099-8e30-f6a9914accae)

yubiuser committed:
Don't trigger double login on paste event to totp field

Signed-off-by: Christian König <ckoenig@posteo.de>

yubiuser commented:
The issue observed was triggered by `pasting` via Ctl+V  being recognized as two events (`paste` and `keyup`). I pushed a commit to only trigger on `input`. I tested it with pre-filled password with: entering TOTP digit by digit, Ctr-V and pasting with mouse. No double login was performed.

yubiuser commented:
Is there a reason we prevent a TOTP re-use within 30 seconds?

```
2023-11-05 23:13:25.448 [399838/T404370] INFO: 2FA code verified successfully at 0
2023-11-05 23:13:33.616 [399838/T404372] WARNING: 2FA code has already been used (-1, 437220), please wait 27 seconds
2023-11-05 23:13:33.618 [399838/T404372] WARNING: API: Invalid 2FA token
```

This seems to be a too long lock interval.

DL6ER commented:
> Is there a reason we prevent a TOTP re-use within 30 seconds?

It is a Time-Based ***One-Time Password*** Algorithm :-)

FTL has the answer to this question here:

> Verify code
> RFC 6238 (section 4.2): If the calculated value matches the value provided by the user, then the user is authenticated
> RFC 6238 (section 4.3): The server MUST NOT accept a TOTP value generated more than 30 seconds in the future
> RFC 6238 (section 4.3): The server MUST NOT accept a TOTP value generated more than 30 seconds in the past
> RFC 6238 (section 4.3): **The server MUST NOT accept a TOTP value it accepted previously**

furthermore:

> We RECOMMEND a default time-step size of 30 seconds.  This default value of 30 seconds is selected as a balance between security and usability.

and sections 5.3:

> Note that a prover may send the same OTP inside a given time-step window multiple times to a verifier.  The verifier **MUST NOT** accept the second attempt of the OTP after the successful validation has been issued for the first OTP, which ensures one-time only use of an OTP.


https://github.com/pi-hole/FTL/blob/afb6f865187099557aa6bb3332286ca2bf6730e7/src/api/2fa.c#L223-L224

The point is: Someone can somehow see your keystrokes. You enter your password. You enter your TOTP code. They immediately try to use them to log in and they'd succeed. The motivation behind the last point is that, even though they're still within the 30 second window for that code, it shouldn't work for they - specifically because you already used it.

DL6ER committed:
Show error message from FTL if applicable

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
With the most recent changes, the login page can much better react on errors:

## Initial state

![Screenshot from 2023-11-06 11-05-15](https://github.com/pi-hole/web/assets/16748619/62f0f0c0-8048-4dac-b85b-5c5c168391bb)

## No password is supplied, submission is prevented

![Screenshot from 2023-11-06 11-06-54](https://github.com/pi-hole/web/assets/16748619/2ffbe685-92d6-4d45-b27d-fcffea03652c)

## Wrong password is entered

![Screenshot from 2023-11-06 11-05-28](https://github.com/pi-hole/web/assets/16748619/37598a9d-a49e-47dc-9ab1-57a5242aed34)

## Rate-limited
(simply mash <kbd>Enter</kbd>)
![Screenshot from 2023-11-06 11-06-25](https://github.com/pi-hole/web/assets/16748619/fceb49f3-7074-4125-88f2-370cbe800982)

## Wrong 2FA token is provided

![Screenshot from 2023-11-06 11-07-37](https://github.com/pi-hole/web/assets/16748619/1caf1164-7141-4fa2-a235-2300d2faec68)

## Number of API seats is exceeded
(set `webserver.api.max_sessions = 1` and try to login in a private window)
![Screenshot from 2023-11-06 11-09-14](https://github.com/pi-hole/web/assets/16748619/303b6a9f-8652-45ee-bbeb-54ca452fdb3a)

---

For rate-limiting and API seat exhaustion to be reported correctly, you have to use FTL from https://github.com/pi-hole/FTL/pull/1733 (branch `new/num_sessions`)

DL6ER committed:
Show FTL hint if available

Signed-off-by: DL6ER <dl6er@dl6er.de>

DL6ER commented:
The last commit adds the inclusion of `hint`s from FTL (if available). For example:

![Screenshot from 2023-11-06 11-34-05](https://github.com/pi-hole/web/assets/16748619/2f3a047e-e8f1-48d7-ad32-fcacbbb10af9)

and

![Screenshot from 2023-11-06 11-34-21](https://github.com/pi-hole/web/assets/16748619/6fe34b60-0f38-4842-9766-5c15f1166d59)

rdwebdesign commented:
I noticed this wasn't reviewed after your last 2 commits.

I tested and it looks fine, but I'm requesting _yubiuser_ approval, because you 2 were discussing about error/warning messages.

sefinek commented:
With the [LastPass](https://www.lastpass.com) extension, the `width` of the `input` is other.

https://github.com/pi-hole/web/assets/92880265/6d99823e-1291-4ea9-906c-01fa7f7757bb

rdwebdesign commented:
This is an external bug.

LastPass extension modifies HTML page content (adding a `<div>` element to hold the icon) and their modification doesn't work here.
Not our fault, they just make assumptions they shouldn't make.

For example: 
if they had used `position: absolute` instead of relative, the layout wouldn't break here.
![lastpass_issue](https://github.com/pi-hole/web/assets/1385443/36648207-96c9-4e77-9b0f-ebd7607f1def)

sefinek commented:
Maybe it's worth informing them about it?

yubiuser commented:
Sure, go ahead. You can link the issue/analysis here so they know what you are referring to.
